framework/src/utils/RankTwoTensor.C

```diff
@@ -889,6 +927,182 @@ RankTwoTensorTempl<T>::symmetricEigenvaluesEigenvectors(std::vector<T> & eigvals
+  {
+    c = a >= 0.0 ? 1.0 : -1.0;
+    s = 0.0;
+  }
+  else if (std::abs(a) > std::abs(b))
+  {
+    T t = b / a;
+    Real sgn = a >= 0.0 ? 1.0 : -1.0;
+    T u = sgn * std::sqrt(1.0 + t * t);
+    c = 1.0 / u;
+    s = c * t;
+  }
+  else
+  {
+    T t = a / b;
+    Real sgn = b >= 0.0 ? 1.0 : -1.0;
+    T u = sgn * std::sqrt(1.0 + t * t);
+    s = 1.0 / u;
+    c = s * t;
+  }
+
+  RankTwoTensorTempl<T> R(initIdentity);
+  R(row1, row1) = c;
+  R(row1, row2) = s;
+  R(row2, row1) = -s;
+  R(row2, row2) = c;
+
+  return R;
+}
+
+template <typename T>
+void
+RankTwoTensorTempl<T>::hessenberg(RankTwoTensorTempl<T> & H, RankTwoTensorTempl<T> & U) const
+{
+  H = *this;
+  U.zero();
+  U.addIa(1.0);
+
+  if (N < 3)
+    return;
+
+  RankTwoTensorTempl<T> R = this->givensRotation(N - 2, N - 1, 0);
+  H = R * H * R.transpose();
+  U = U * R.transpose();
+}
+
+template <typename T>
+void
+RankTwoTensorTempl<T>::QR(RankTwoTensorTempl<T> & Q,
+                          RankTwoTensorTempl<T> & R,
+                          unsigned int dim) const
+{
+  R = *this;
+  Q.zero();
+  Q.addIa(1.0);
+
+  for (unsigned int i = 0; i < dim - 1; i++)
+    for (unsigned int b = dim - 1; b > i; b--)
+    {
+      unsigned int a = b - 1;
+      RankTwoTensorTempl<T> CS = R.givensRotation(a, b, i);
+      R = CS * R;
+      Q = Q * CS.transpose();
+    }
+}
+
+template <>
+void
+RankTwoTensorTempl<DualReal>::QR(RankTwoTensorTempl<DualReal> & Q,
+                                 RankTwoTensorTempl<DualReal> & R,
+                                 unsigned int dim) const
+{
+  R = *this;
+  Q.zero();
+  Q.addIa(1.0);
+
+  for (unsigned int i = 0; i < dim - 1; i++)
+    for (unsigned int b = dim - 1; b > i; b--)
+    {
+      unsigned int a = b - 1;
+
+      // special case when both entries to rotate are zero
+      // in which case the dual numbers cannot be rotated
+      // therefore we need to find another nonzero entry to permute
+      RankTwoTensorTempl<DualReal> P(initIdentity);
+      if (MooseUtils::absoluteFuzzyEqual(R(a, i).value(), 0.0) &&
+          MooseUtils::absoluteFuzzyEqual(R(b, i).value(), 0.0))
+      {
+        unsigned int c = 3 - a - b;
+        if (!MooseUtils::absoluteFuzzyEqual(R(c, i).value(), 0.0))
+          P = this->permutationTensor({a, b, c}, {c, b, a});
+      }
+
+      Q = Q * P.transpose();
+      R = P * R;
+      RankTwoTensorTempl<DualReal> CS = R.givensRotation(a, b, i);
+      R = P.transpose() * CS * R;
+      Q = Q * CS.transpose() * P;
+    }
+}
+
+template <>
+void
+RankTwoTensorTempl<DualReal>::symmetricEigenvaluesEigenvectors(
+    std::vector<DualReal> & eigvals, RankTwoTensorTempl<DualReal> & eigvecs) const
+{
+  const Real eps = libMesh::TOLERANCE * libMesh::TOLERANCE;
+
+  eigvals.resize(N);
+  RankTwoTensorTempl<DualReal> D, Q, R;
+  this->hessenberg(D, eigvecs);
+
+  unsigned int iter = 0;
+  for (unsigned m = N - 1; m > 0; m--)
```

Comment:
How did this not error? unsigned ints are always greater than zero. This is a bad loop test!

Reply:
Hmm I learned it a long time ago so I might be mistaken -- isn't 000...0 in signed bit a zero?

Reply:
Sorry - you are correct

` m > 0` is valid
` m >= 0` is not valid for unsigned